Update and rename 2022-06-08-redis.md to 2023-06-27-redis.md (#10)
bubunyo/bubunyo.github.io: _posts/2023-06-27-redis.md

---
layout: post
title: "Redis"
description: "Why is redis fast, what are the pros and cons of using redis"
comments: true
keywords: "redis, infrastructure"
published: false
---

> [Redis](https://redis.io) was designed to be a blazingly fast key-value store to make caching easy. And for most parts, it lives up to the billing. Setting up and using Redis has always been a breeze. The success and popularity of Redis has made it a very powerful tool in the software engineers' arsenal; with its support of almost all data types and associated operations out of the box, it is a perfect solution for almost all persistence problems. As with everything, people have used it in ways without considering its fitness for the problem at hand, often to great success.

Reviewer: Not to be pendantic but this is an important clarificaiton. I think if you listen to Antirez, he appears to not have written redis to cache data per se. His writings about why he wrote it implies he wrote redis because he wanted a fast in-memory data structure server so the more correct way to describe redis is `a key-value in-memory database that trades some durability for speed`. That it's used commonly for caching is a side-effect of its speed.

You can see clearly how redis has progressed over time confirms that. It's rich in data structures from common ones to esoteric ones like hyperloglog and bloom filters. It even has full blown distributed locking support and data streaming primitives now.

> [Redis](https://redis.io) wass designed to be a blazingly fast key-value store to make caching easy. And for most parts it lives up to the billing. Setting up and using redis has always been a breeze. The success and popularity of redis has makes it it a very powerful tool in the software engineers arsenal, with its support of almost all data type and associated operations out of the box, it is a perfect solution for all persistence problems; almost. As with everything, people have used it in ways without consideration for if its fitness for the problem at hand often to great success.

Reviewer 2: ```suggestion
[Redis](https://redis.io) was designed to be a blazingly fast key-value store to make caching easy. And for most parts, it lives up to the billing. Setting up and using Redis has always been a breeze. The success and popularity of Redis has made it a very powerful tool in the software engineers' arsenal; with its support of almost all data types and associated operations out of the box, it is a perfect solution for almost all persistence problems. As with everything, people have used it in ways without considering its fitness for the problem at hand, often to great success.
```

Redis is famed for being able to perform operations that is measured in sub milliseconds. And the key to this is how it is designed. 

> It is designed to store data in memory.  If your first year computer science class is anything to go by, you will know this provides much fast read and write throughput  compared to disk by miles. When we create data structures using variables we often use arrays, hashes and maps. Redis supports these and more. it is also effecient and manipulating them because, It takes advantage of the underlying data structures that are optimized for in memory storage without worrying about how to persist it to durable storage. .It is also single threaded, Though a performant single thread system might sound counter intuitive, that are some peculiar performant advantages to it. And redis takes advantage of this in a brilliant way to ensure consistency without an cost to performance. Redis's single-thread will scales indefinitely in terms of I/O concurrency. It does this by using an I/O demultiplexing mechanism and a concise event loop designed by the author. Thus there is no synchronization to be done since all commands are serialized. It might look like CPU might become a bottle neck with this design but it turns out more often than not, you will hit a network bottleneck well before CPU cannot keep up. The positive side effect of this design is that atomicity of all operations come at no extra cost. Redis also use a [proprietary protocol](https://redis.io/docs/reference/protocol-spec/), that is much more terse.  Couple the isolated event loop with a proprietary communication protocol and you have a blazing fast in memory data store that scales indefinitely in theory. 

Author: ```suggestion
It is designed to store data in memory.  If your first year computer science class is anything to go by, you will know this provides much fast read and write throughput  compared to disk by miles. When we create data structures using variables we often use arrays, hashes and maps. Redis supports these and more. it is also effecient and manipulating them because, It takes advantage of the underlying data structures that are optimized for in memory storage without worrying about how to persist it to durable storage. It is also single threaded, Though a performant single thread system might sound counter intuitive, that are some peculiar performant advantages to it. And redis takes advantage of this in a brilliant way to ensure consistency without any cost to performance. Redis's single-thread will scales indefinitely in terms of I/O concurrency. It does this by using an I/O demultiplexing mechanism and a concise event loop designed by the author. Thus there is no synchronization to be done since all commands are serialized. It might look like CPU might become a bottle neck with this design but it turns out more often than not, you will hit a network bottleneck well before CPU cannot keep up. The positive side effect of this design is that atomicity of all operations come at no extra cost. Redis also use a [proprietary protocol](https://redis.io/docs/reference/protocol-spec/), that is much more terse.  Couple the isolated event loop with a proprietary communication protocol and you have a blazing fast in memory data store that scales indefinitely in theory and you have a highly preformant database. 
```

> It is designed to store data in memory.  If your first year computer science class is anything to go by, you will know this provides much fast read and write throughput  compared to disk by miles. When we create data structures using variables we often use arrays, hashes and maps. Redis supports these and more. it is also effecient and manipulating them because, It takes advantage of the underlying data structures that are optimized for in memory storage without worrying about how to persist it to durable storage. It is also single threaded, Though a performant single thread system might sound counter intuitive, that are some peculiar performant advantages to it. And redis takes advantage of this in a brilliant way to ensure consistency without any cost to performance. Redis's single-thread will scales indefinitely in terms of I/O concurrency. It does this by using an I/O demultiplexing mechanism and a concise event loop designed by the author. Thus there is no synchronization to be done since all commands are serialized. It might look like CPU might become a bottle neck with this design but it turns out more often than not, you will hit a network bottleneck well before CPU cannot keep up. The positive side effect of this design is that atomicity of all operations come at no extra cost. Redis also use a [proprietary protocol](https://redis.io/docs/reference/protocol-spec/), that is much more terse.  Couple the isolated event loop with a proprietary communication protocol and you have a blazing fast in memory data store that scales indefinitely in theory and you have a highly preformant database. 

Reviewer 2: ```suggestion
It is designed to store data in memory.  If your first-year computer science class is anything to go by, you will know this provides much faster read and write throughput compared to disk by miles. When we create data structures using variables, we often use arrays, hashes and maps. Redis supports these and more. It is also efficient at manipulating them because it takes advantage of the underlying data structures that are optimized for in-memory storage without worrying about how to persist it to durable storage. It is also single-threaded, Though a performant single-thread system might sound counterintuitive, there are some peculiar performant advantages to it. And Redis takes advantage of this in a brilliant way to ensure consistency without any cost to performance. Redis's single-thread will scale indefinitely in terms of I/O concurrency. It does this by using an I/O demultiplexing mechanism and a concise event loop designed by the author. Thus there is no synchronization to be done since all commands are serialized. It might look like the CPU might become a bottleneck with this design, but it turns out you will often hit a network bottleneck well before the CPU cannot keep up. The positive side effect of this design is that the atomicity of all operations comes at no extra cost. Redis also uses a [proprietary protocol](https://redis.io/docs/reference/protocol-spec/) that is much more terse. Couple the isolated event loop with a proprietary communication protocol, and you have a blazing fast in-memory data store that scales indefinitely in theory, and you have a highly performant database. 
```

> These facts only holds when the size of your payload and the number of connections remain relatively small. this easily jumps out the window with ever increasing load parameters. The threshold is unfortunately rather low at a high number connections and increased payload sizes. Modern large scale micro-services will easily have over 100 running instance at medium scale. And since most instances employing some pooling mechanism to so as not to pay a connection cost for each request, A single redis instance is going to a bit of work in maintaining those connections not to talk of serving requests as they come through. To improve performance at medium to high loads, some projects such as KeyDb which is supposed to be a drop in replacement for redis employs multithreaded approach and a bit of magic to sustain some high workloads. This has been touted to provide 5x performance over redis. Another solution which I have seen used is to employ a proxy that multiplexes over multiple redises. One such proxy is the twoemproxy developed at twitter. [Twoemproxy](https://github.com/twitter/twemproxy) is in itself single threaded and employs key hashing to store keys in shards of multiple redis instance give you proper multiplexing. While it may look like this is susceptible to the original problems of a single threaded application, it is not necessarily the reality, because twoemproxy employs a single thread for each redis instance, turning the whole system into a multi threaded systems. Of course this is still susceptible to hotkeys. These might look like ideal solutions, setting up new infrastructure as an intermediary service introduces a new failure point and this is neither trivial nor ideal. But when done, there is a lot of net positives. 

Author: ```suggestion
These facts only holds when the size of your payload and the number of connections remain relatively small. this easily jumps out the window with ever increasing load parameters. The threshold is unfortunately rather low at a high number connections and increased payload sizes. Modern large scale micro-services will easily have over 100 running instance at medium scale. And since most instances employing some pooling mechanism so as not to pay a connection cost for each request, A single redis instance is going to do a bit of work in maintaining those connections not to talk of serving requests as they come through. To improve performance at medium to high loads, some projects such as [KeyDb](https://docs.keydb.dev), Snapchats drop-in replacement as an alternative to redis, employs multithreaded approach and a bit of magic to sustain some high workloads. This has been touted to provide 5x performance over redis. Another solution which I have seen used is to employ a proxy that multiplexes over multiple redises. One such proxy is the twoemproxy developed at twitter. [Twoemproxy](https://github.com/twitter/twemproxy), or nutcracker as it is informally know, is in itself single threaded and employs key hashing to store keys in shards of multiple redis instance give you proper multiplexing. While it may look like this is susceptible to the original problems of a single threaded application, it is not necessarily the reality, because twoemproxy employs a single thread for each redis instance, turning the whole system into a multi threaded systems. Of course this is still susceptible to hotkeys. These might look like ideal solutions, setting up new infrastructure as an intermediary service introduces a new failure point and this is neither trivial nor ideal. But when done, there is a lot of net positives. 
```

> These facts only holds when the size of your payload and the number of connections remain relatively small. this easily jumps out the window with ever increasing load parameters. The threshold is unfortunately rather low at a high number connections and increased payload sizes. Modern large scale micro-services will easily have over 100 running instance at medium scale. And since most instances employing some pooling mechanism so as not to pay a connection cost for each request, A single redis instance is going to do a bit of work in maintaining those connections not to talk of serving requests as they come through. To improve performance at medium to high loads, some projects such as [KeyDb](https://docs.keydb.dev), Snapchats drop-in replacement as an alternative to redis, employs multithreaded approach and a bit of magic to sustain some high workloads. This has been touted to provide 5x performance over redis. Another solution which I have seen used is to employ a proxy that multiplexes over multiple redises. One such proxy is the twoemproxy developed at twitter. [Twoemproxy](https://github.com/twitter/twemproxy), or nutcracker as it is informally know, is in itself single threaded and employs key hashing to store keys in shards of multiple redis instance give you proper multiplexing. While it may look like this is susceptible to the original problems of a single threaded application, it is not necessarily the reality, because twoemproxy employs a single thread for each redis instance, turning the whole system into a multi threaded systems. Of course this is still susceptible to hotkeys. These might look like ideal solutions, setting up new infrastructure as an intermediary service introduces a new failure point and this is neither trivial nor ideal. But when done, there is a lot of net positives. 

Reviewer 2: ```suggestion
These facts only hold when the size of your payload and the number of connections remain relatively small. This easily jumps out the window with ever-increasing load parameters. The threshold is, unfortunately, rather low at a high number of connections and increased payload sizes. Modern large-scale micro-services will easily have over 100 running instances at medium scale. And since most instances employ some pooling mechanism so as not to pay a connection cost for each request, a single Redis instance is going to do a bit of work in maintaining those connections, not to talk of serving requests as they come through. To improve performance at medium to high loads, some projects such as [KeyDb](https://docs.keydb.dev), Snapchat's drop-in replacement as an alternative to Redis, employ a multithreaded approach and a bit of magic to sustain some high workloads. This has been touted to provide 5x performance over Redis. Another solution which I have seen used is to employ a proxy that multiplexes over multiple redises. One such proxy is the twoemproxy developed at Twitter. [Twoemproxy](https://github.com/twitter/twemproxy), or nutcracker as it is informally known, is in itself single-threaded and employs key hashing to store keys in shards of multiple redis instances, giving you proper multiplexing. While it may look like this is susceptible to the original problems of a single-threaded application, it is not necessarily the reality because twoemproxy employs a single thread for each Redis instance, turning the whole system into a multi-threaded system. Of course, this is still susceptible to hotkeys. These might look like ideal solutions, but setting up new infrastructure as an intermediary service introduces a new failure point, which is neither trivial nor ideal. But when done right, there are a lot of net positives. 
```

> These facts only hold when the size of your payload and the number of connections remain relatively small. This easily jumps out the window with ever-increasing load parameters. The threshold is, unfortunately, rather low at a high number of connections and increased payload sizes. Modern large-scale micro-services will easily have over 100 running instances at medium scale. And since most instances employ some pooling mechanism so as not to pay a connection cost for each request, a single Redis instance is going to do a bit of work in maintaining those connections, not to talk of serving requests as they come through. To improve performance at medium to high loads, some projects such as [KeyDb](https://docs.keydb.dev), Snapchat's drop-in replacement as an alternative to Redis, employ a multithreaded approach and a bit of magic to sustain some high workloads. This has been touted to provide 5x performance over Redis. Another solution which I have seen used is to employ a proxy that multiplexes over multiple redises. One such proxy is the twoemproxy developed at Twitter. [Twoemproxy](https://github.com/twitter/twemproxy), or nutcracker as it is informally known, is in itself single-threaded and employs key hashing to store keys in shards of multiple redis instances, giving you proper multiplexing. While it may look like this is susceptible to the original problems of a single-threaded application, it is not necessarily the reality because twoemproxy employs a single thread for each Redis instance, turning the whole system into a multi-threaded system. Of course, this is still susceptible to hotkeys. These might look like ideal solutions, but setting up new infrastructure as an intermediary service introduces a new failure point, which is neither trivial nor ideal. But when done right, there are a lot of net positives. 

Reviewer: Good overall but you might want to talk about [redis-cluster](https://redis.io/docs/management/scaling/) or [master slave](https://redis.io/docs/management/replication/) coupled with [redis-sentinel](https://redis.io/docs/management/sentinel/) as  a way to horizontally scale redis deployments and its downsides.

Author: i delibrately avoided them because people hardly use them in the realworld and durability is not guaranteed even in cluster mode

> Another major concern with redis is durability. Redis out of the box does not persist data on disk, only in memory. This means, when a server goes down, so does all you data. Durability is serious business and when it becomes a priority. And this is where redis starts to go backwards. Redis was never planned to be provide durable beyond RAM. This is evident in the fact that disk persistence, was never part of redis until, v 0.04. Redis supports two types of [persistence modes](https://redis.io/docs/management/persistence/).

Author: ```suggestion
Another major concern with redis is durability. Redis out of the box does not persist data on disk, only in memory. This means, when a server goes down, so does all you data. Durability is serious business and when it becomes a priority. And this is where redis starts to go backwards. Redis was never planned to be provide durability beyond RAM. This is evident in the fact that disk persistence, was never part of redis until, v 0.04. Redis supports two types of [persistence modes](https://redis.io/docs/management/persistence/).
```

> Another major concern with Redis is durability. Redis out of the box does not persist data on disk, only in memory. This means when a server goes down, so does all your data. Durability is serious business and when it becomes a priority, this is where Redis starts to go backwards. Redis was never planned to provide durability beyond RAM. This is evident in the fact that disk persistence was never part of Redis until v0.04. Redis supports two types of [persistence modes](https://redis.io/docs/management/persistence/).

Reviewer: Out of the box I think it does persist data. By default it's set to rdb persistence. For example [redis 7.0 has the following as default](https://raw.githubusercontent.com/redis/redis/7.0/redis.conf)

```
#
# Unless specified otherwise, by default Redis will save the DB:
#   * After 3600 seconds (an hour) if at least 1 change was performed
#   * After 300 seconds (5 minutes) if at least 100 changes were performed
#   * After 60 seconds if at least 10000 changes were performed
#
```

See the configurations [here](https://redis.io/docs/management/config/)

Author: i think i did mention this down the line.

> Another major concern with redis is durability. Redis out of the box does not persist data on disk, only in memory. This means, when a server goes down, so does all you data. Durability is serious business and when it becomes a priority. And this is where redis starts to go backwards. Redis was never planned to be provide durability beyond RAM. This is evident in the fact that disk persistence, was never part of redis until, v 0.04. Redis supports two types of [persistence modes](https://redis.io/docs/management/persistence/).

Reviewer 2: ```suggestion
Another major concern with Redis is durability. Redis out of the box does not persist data on disk, only in memory. This means when a server goes down, so does all your data. Durability is serious business and when it becomes a priority, this is where Redis starts to go backwards. Redis was never planned to provide durability beyond RAM. This is evident in the fact that disk persistence was never part of Redis until v0.04. Redis supports two types of [persistence modes](https://redis.io/docs/management/persistence/).
```

> The first one is called snapshotting which Redis calls RDB. When snapshotting is enabled redis will periodically write all your dataset in memory to disk. This is good for point in time recovery. But this also means, you loose all the data between when the last snapshot was created versus when the failure occurred. For a moderately busy server there are bound to be significant changes that happen between when snapshots are set, and loosing it might not be a good thing. To combat this, some teams, set the snapshots times as minimal as possible to combat the amount of data it is possible to loose. This can be a bad idea, when your data set is considerably large. Writing a 1gig file to disk every every 60 seconds, is a recipe for disaster. [What happens in the background](https://redis.io/docs/getting-started/faq/#background-saving-fails-with-a-fork-error-on-linux) is redis forks a child process for background processing, serialize in memory and make it disk compatible, write it to a temporary file in the background and rename the file atomically to  one upon finish. Even though the overhead of creating a fork is zero in theory when the OS supports copy-on-write, you still need to turn on the `overcommit_memory`. This is because if the dataset between the parent process and the child process deviate, redis will not keep track of the changes and will have to allocate just as much memory your data set has to the child process in order to snapshot successful. In snapshotting mode, you must do everything to ensure that, your datastore does not exceed half the RAM allocated to redis, otherwise your redis server will implode with an OOM. This is the default persistence mode cos it is simple and safe for small data sets. When you data set starts to increase considerably in size, think twice about snapshots. 

Author: ```suggestion
The first one is called snapshotting or RDB. When snapshotting is enabled redis will periodically write all your dataset in memory to disk. This is good for point in time recovery. But this also means, you loose all the data between when the last snapshot was created versus when the failure occurred. For a moderately busy server there are bound to be significant changes that happen between when snapshots are set, and loosing it might not be a good thing. To combat this, some teams, set the snapshots times as minimal as possible to combat the amount of data it is possible to loose. This can be a bad idea, when your data set is considerably large. Writing a 1gig file to disk every every 60 seconds, is a recipe for disaster. [What happens in the background](https://redis.io/docs/getting-started/faq/#background-saving-fails-with-a-fork-error-on-linux) is redis forks a child process for background processing, serializing the dataset in memory, making it disk compatible, writing it to a temporary file in the background and rename the file atomically upon finish. Even though the overhead of creating a fork is zero in theory when the OS supports copy-on-write, you still need to turn on the `overcommit_memory`. This is because if the dataset between the parent process and the child process deviate, redis will not keep track of the changes and will have to allocate just as much memory your data set has to the child process in order for the snapshot to be successful. In snapshotting mode, you must do everything to ensure that, your datastore does not exceed half the RAM allocated to redis, otherwise your redis server will implode with an OOM. This is the default persistence mode cos it is simple and safe for small data sets. When you data set starts to increase considerably in size, think twice about snapshots. 
```

> The first one is called snapshotting or RDB. When snapshotting is enabled, Redis will periodically write all your dataset in memory to disk. This is good for point-in-time recovery. But this also means you lose all the data between when the last snapshot was created versus when the failure occurred. For a moderately busy server, there are bound to be significant changes that happen between when snapshots are set, and losing it might not be a good thing. To combat this, some teams set the snapshot times as minimal as possible to combat the amount of data it is possible to lose. This can be a bad idea when your data set is considerably large. Writing a 1gig file to disk every 60 seconds is a recipe for disaster. [What happens in the background](https://redis.io/docs/getting-started/faq/#background-saving-fails-with-a-fork-error-on-linux) is Redis forks a child process for background processing, serializing the dataset in memory, making it disk compatible, writing it to a temporary file in the background and rename the file atomically upon finish. Even though the overhead of creating a fork is zero in theory when the OS supports copy-on-write, you still need to turn on the `overcommit_memory`. This is because if the dataset between the parent process and the child process deviates, Redis will not keep track of the changes and will have to allocate just as much memory your data set has to the child process in order for the snapshot to be successful. In snapshotting mode, you must do everything to ensure that your datastore does not exceed half the RAM allocated to Redis, otherwise, your Redis server will implode with an OOM. This is the default persistence mode because it is simple and safe for small data sets. When your data set starts to increase considerably in size, think twice about snapshots. 

Reviewer: Great over all but you might also want to add the other parameter for rdb: number of writes. So you have two variables, at least N seconds or M writes. Which ever comes first.

> The first one is called snapshotting or RDB. When snapshotting is enabled redis will periodically write all your dataset in memory to disk. This is good for point in time recovery. But this also means, you loose all the data between when the last snapshot was created versus when the failure occurred. For a moderately busy server there are bound to be significant changes that happen between when snapshots are set, and loosing it might not be a good thing. To combat this, some teams, set the snapshots times as minimal as possible to combat the amount of data it is possible to loose. This can be a bad idea, when your data set is considerably large. Writing a 1gig file to disk every every 60 seconds, is a recipe for disaster. [What happens in the background](https://redis.io/docs/getting-started/faq/#background-saving-fails-with-a-fork-error-on-linux) is redis forks a child process for background processing, serializing the dataset in memory, making it disk compatible, writing it to a temporary file in the background and rename the file atomically upon finish. Even though the overhead of creating a fork is zero in theory when the OS supports copy-on-write, you still need to turn on the `overcommit_memory`. This is because if the dataset between the parent process and the child process deviate, redis will not keep track of the changes and will have to allocate just as much memory your data set has to the child process in order for the snapshot to be successful. In snapshotting mode, you must do everything to ensure that, your datastore does not exceed half the RAM allocated to redis, otherwise your redis server will implode with an OOM. This is the default persistence mode cos it is simple and safe for small data sets. When you data set starts to increase considerably in size, think twice about snapshots. 

Reviewer 2: ```suggestion
The first one is called snapshotting or RDB. When snapshotting is enabled, Redis will periodically write all your dataset in memory to disk. This is good for point-in-time recovery. But this also means you lose all the data between when the last snapshot was created versus when the failure occurred. For a moderately busy server, there are bound to be significant changes that happen between when snapshots are set, and losing it might not be a good thing. To combat this, some teams set the snapshot times as minimal as possible to combat the amount of data it is possible to lose. This can be a bad idea when your data set is considerably large. Writing a 1gig file to disk every 60 seconds is a recipe for disaster. [What happens in the background](https://redis.io/docs/getting-started/faq/#background-saving-fails-with-a-fork-error-on-linux) is Redis forks a child process for background processing, serializing the dataset in memory, making it disk compatible, writing it to a temporary file in the background and rename the file atomically upon finish. Even though the overhead of creating a fork is zero in theory when the OS supports copy-on-write, you still need to turn on the `overcommit_memory`. This is because if the dataset between the parent process and the child process deviates, Redis will not keep track of the changes and will have to allocate just as much memory your data set has to the child process in order for the snapshot to be successful. In snapshotting mode, you must do everything to ensure that your datastore does not exceed half the RAM allocated to Redis, otherwise, your Redis server will implode with an OOM. This is the default persistence mode because it is simple and safe for small data sets. When your data set starts to increase considerably in size, think twice about snapshots. 
```

> The second more durable persistence mode is Append Only File which was introduced in Redis 1.1 to solve the drawbacks of snapshotting. In this mode, every redis command, is appended to a file as a log. Very much like WAL logs for conventional RDBMSs. This way, you can build the entire database  by replaying the entire file. If can be argued that sequential writes to file are significantly faster than random access but is is still significantly slower than writing in memory, and this goes against the essence of what redis is, which is an in-memory data store. If you are going to be writing each command to a file, why not use a datastore that is designed for that in the first place. If you use AOF it means redis is going to call fsync at a point in time which can be configured in 1 of 3 ways. `appendfsync always` will call fsync on each command. This is very very safe. But with this option you might as well through redis, out the window, because the performance becomes infirior to every database designed to fsync on write. if you use this mode without a [Battery Backed Write Cache (BBWC) RAID controller](https://serverfault.com/questions/65096/battery-backed-write-cache), you will get fucked. Have fun figuring out what went wrong. `appendfsync everysec` will call `fsync` every second which means at most you loose a second of data. This might sound reasonable but is not without its drawbacks. if you have a update intensive application like a counter that updates many times in an second, you end up with a needlessly huge AOF file when in essence the data payload might only have a small foot print. that said is is an easier choice and has been the new default since Redis 2.4. `appendfsync no` delegates the calling of `fsync` to the operating system. This is the fastest and the least safe method amongst all the `appendfsync` options. Normally Linux will flush data every 30 seconds with this configuration, but it's up to the kernel's exact tuning. 

Reviewer 2: ```suggestion
The second more durable persistence mode is Append Only File, introduced in Redis 1.1 to solve the drawbacks of snapshotting. In this mode, every Redis command is appended to a file as a log. Very much like WAL logs for conventional RDBMSs. This way, you can build the entire database by replaying the entire file. It can be argued that sequential writes to files are significantly faster than random access but is is still significantly slower than writing in memory, and this goes against the essence of what Redis is, which is an in-memory data store. If you are going to be writing each command to a file, why not use a datastore that is designed for that in the first place? If you use AOF, it means Redis is going to call fsync at a point in time which can be configured in 1 of 3 ways. `appendfsync always` will call fsync on each command. This is very very safe. But with this option, you might as well throw Redis out the window because the performance becomes inferior to every database designed to fsync on write. If you use this mode without a [Battery Backed Write Cache (BBWC) RAID controller](https://serverfault.com/questions/65096/battery-backed-write-cache), you will get fucked. Have fun figuring out what went wrong. `appendfsync everysec` will call `fsync` every second, which means, at most you lose a second of data. This might sound reasonable, but it is not without its drawbacks. If you have an update-intensive application like a counter that updates many times in a second, you end up with a needlessly huge AOF file when in essence, the data payload might only have a small footprint. That said, it is an easier choice and has been the new default since Redis 2.4. `appendfsync no` delegates the calling of `fsync` to the operating system. This is the fastest and the least safe method amongst all the `appendfsync` options. Normally Linux will flush data every 30 seconds with this configuration, but it's up to the kernel's exact tuning. 
```

> When i talk to other developers about systems design problems, most people are quick to suggest redis as a solution to buffer data when producers can not keep up with consumers, leaning heavily on the high write throughput of redis. The solution often goes like this; use redis as a cache layer between the two entities. While this design generally works, I often wonder if people truly understand the tradeoffs being made here and the risks involved. The lack of proper durability in redis make the bandwidth of solutions it is a fit for a very narrow one, nonetheless for that bandwidth it offers tremendous advantages. If you are going to drop redis into you stack, be sure it is absolutely what you need, or stay well clear of it. Otherwise you are you data are going to get burnt. 

Reviewer 2: ```suggestion
When I talk to other developers about systems design problems, most people are quick to suggest Redis as a solution to buffer data when producers can not keep up with consumers, leaning heavily on the high write-throughput of Redis. The solution often goes like this; use Redis as a cache layer between the two entities. While this design generally works, I often wonder if people truly understand the tradeoffs and risks involved here. The lack of proper durability in Redis makes the bandwidth of solutions it is a fit for a very narrow one; nonetheless, for that bandwidth, it offers tremendous advantages. If you will drop Redis into your stack, be sure it is absolutely what you need, or stay clear of it. Otherwise, you and your data are going to get burnt. 
```
